### PR TITLE
Export trained model into a Pipeline object

### PR DESCRIPTION
**Pull request by exviva**
This way we can use the trained model for predictions. With serialization, we can also restore it later on, even in a different environment (e.g. xgboost4j).

### CODE REVIEW

File: guacml/enums.py [hunk 1]
```diff
@@ -1,6 +0,0 @@
-from enum import Enum
-
-ProblemType = Enum('ProblemType', 'BINARY_CLAS MULTI_CLAS REGRESSION')
```
Comment (hwartig) [inline guacml/enums.py:3]: I don't understand why you removed the enums. To me using something typed is preferable over using plain strings and I think python has also [support for pickling enums](https://docs.python.org/3/library/enum.html#pickling). Is this to follow some python best practice? Do you mind to elaborate?

File: guacml/target_transforms/transformed_eval_metric.py [hunk 2]
```diff
@@ -1,17 +0,0 @@
-from guacml.metrics.base_eval_metric import BaseEvalMetric
-
-
-class TransformedEvalMetric(BaseEvalMetric):
```
Comment (hwartig) [inline guacml/target_transforms/transformed_eval_metric.py:4]: Nice catch 👍 for cleanup

File: guacml/models/linear_model.py [hunk 3]
```diff
@@ -1,38 +1,40 @@
...
 from guacml.models.base_model import BaseModel
-from guacml.preprocessing.column_analyzer import ColType
-import pandas as pd
-import numpy as np
 
 
 class LinearModel(BaseModel):
 
+    def copy_model(self):
+        return pickle.loads(pickle.dumps(self.model))
```
Comment (hwartig) [inline guacml/models/linear_model.py:14]: did you consider using [`copy.deep_copy`](https://docs.python.org/3/library/copy.html#copy.deepcopy) or is it intentional that you go through pickle?

File: guacml/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
+from guacml.dataset import Dataset
+from guacml.preprocessing.column_analyzer import ColumnAnalyzer
+
+
+class Pipeline:
+
+    def __init__(self, guac, model_name):
+        mr = guac.model_results[model_name]
```
Comment (hwartig) [inline guacml/pipeline.py]: If im not mistaken `model_results == None` if `guac.run` wasn't called yet. WDYT about raising a nice error to the user to tell him about this dependency? Another option could be to call `guac.run` in `Pipeline.__init__` if id wasn't run already before and only notify the user about the fact that we did.

File: guacml/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+        self.tree = guac.tree.get_subtree_upto(model_name)
+        self.config = self.tree.config
+        self.config['column_types'] = mr.metadata.type.to_dict()
+        self.logger = guac.logger
+        self.model = mr.model.copy(self.config)
+        self.features = list(mr.features)
+
+    def transform(self, test_set):
+        step_name = self.tree.root_name
+        metadata = ColumnAnalyzer(self.config, self.logger).analyze(test_set)
```
Comment (hwartig) [inline guacml/pipeline.py]: I think we should serialize the metadata of the `ColumnAnalyzer` that was used to train the model for the following reasons:

- it seems like an unnecessary overhead to reanalyse the dataset at prediction time (even if its just a small one). For production use cases we want the prediction step to be as fast as possible
- if the dataset that is passed into the prediction step is small (imagine one single row) the results of `analyse` might be vastly different form the `analyse` result that we had during training of the model.
- if the given dataset would rightfully lead to a different set of metadata we should raise an error. Lets say there was a string supplied where we expected an integer. In cases like this we should tell the user that we expected something else in this column. There are cases where it is not so easy though. One that comes to mind is: What if we see a new categorical value on prediction time? Should we ignore it or report an error?

WDYT?

File: guacml/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+        while step_name is not None:
+            step = self.tree.get_step(step_name)
+            data = step.execute(data)
+            children = self.tree.get_children(step_name)
+            step_name = children[0] if len(children) > 0 else None
+
+        return data.df
+
+    def predict(self, test_set):
```
Comment (hwartig) [inline guacml/pipeline.py]: I don't like the naming `test_set` here. WDYT of sticking to the [sklearn convention](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html#sklearn.ensemble.RandomForestClassifier.predict) and naming it `X`?

File: guacml/models/random_forest.py [hunk 7]
```diff
@@ -56,3 +57,6 @@ def predict(self, x):
     def feature_importances(self, x):
         feat_scores = self.model.feature_importances_
         return pd.Series(list(feat_scores), index=list(x.columns))
+
+    def get_state(self):
```
Comment (hwartig) [inline guacml/models/random_forest.py:61]: I wonder if we could move this into the `BaseModel` as the default implementation and only override it when needed. Like for XGBoost.

File: tests/test_pipeline.py [hunk 8]
```diff
@@ -1,19 +1,24 @@
...
+import base64
 
 from guacml.pipeline import Pipeline
 from tests.test_util import load_dataset, read_fixture
 
 
 class TestPipeline(unittest.TestCase):
 
-    def test_predict(self):
+    def quick_run(self):
```
Comment (hwartig) [inline tests/test_pipeline.py]: The `quick_run` would be even quicker if we created a smaller tree to begin with rather then first creating the full thing and then removing steps and models. ;-)

But more importantly should we test the other models as well or do you think focussing only on XGBoost is good enough?

File: guacml/enums.py (hunk 1 above)
Comment (exviva) [reply guacml/enums.py:3]: The main reason was that enums are not JSON-serializable in python, and I needed portability outside of python, so I couldn't use pickle.

File: guacml/models/linear_model.py (hunk 3 above)
Comment (exviva) [reply guacml/models/linear_model.py:14]: sklearn is recommending using pickle: http://scikit-learn.org/stable/modules/model_persistence.html. Not sure if `deep_copy` would've made a full deep memory copy with completely fresh references.

File: guacml/models/random_forest.py (hunk 7 above)
Comment (exviva) [reply guacml/models/random_forest.py:61]: Yeah, I thought about it, but I'm still not sure whether using `pickle` is more sklearn-specific, or should be the default. I'm also still not really sure about this whole `get_state`/`copy_model` API, feels a lot like re-inventing the wheel.

If we implement more models and they do the same thing in `get_state`, we can put it on the base class.

File: guacml/pipeline.py (hunk 6 above)
Comment (exviva) [reply guacml/pipeline.py]: Me neither. I like `x`, will change

**Review by exviva — COMMENTED** [review]
Thanks for the feedback @hwartig , wanna have one more look? I've added some more commits

File: guacml/models/linear_model.py (hunk 3 above)
Comment (hwartig) [reply guacml/models/linear_model.py:14]: got it! pickle it is.

File: guacml/enums.py (hunk 1 above)
Comment (hwartig) [reply guacml/enums.py:3]: Couldn't we write a custom serialiser/deserialiser. This way we could keep the type save python enums and also during deserialisation give out nice error messages for the users that prefer the json way. The tradeoff for me is not even so much about cleaner code but more about having the better error messages for our users so that they don't run into errors just for misspelling the error metric they wanna use. WDYT?

File: guacml/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        self.name = name
+        self.tree = tree
+        self.config = config
+        self.logger = logger
+        self.model = model
+        self.features = features
+
+    def transform(self, x):
+        step_name = self.tree.root_name
+        metadata = ColumnAnalyzer(self.config, self.logger).analyze(x)
```
Comment (hwartig) [inline guacml/pipeline.py:33]: @exviva did you see my [earlier comment](https://github.com/guacml/guacml/pull/18#discussion_r131241134)?

Comment (hwartig) [reply guacml/pipeline.py:33]: I think if we split up column analyzing and column transforming we could have a api where the `ColumnAnalyzer#analyze(train_x)` returns a `ColumnTransformer` that supports `ColumnTransformer#transform(x)` and is serializable at the same time. WDYT?

File: guacml/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+from guacml.preprocessing.column_analyzer import ColumnAnalyzer
+from guacml.step_tree.step_tree import StepTree
+from guacml.util import deep_copy, get_fully_qualified_class_name, get_class_from_string
+
+
+class Pipeline:
+
+    @staticmethod
+    def from_tree(guac, model_name):
+        mr = guac.model_results[model_name]
```
Comment (hwartig) [inline guacml/pipeline.py:13]: Did you see my [earlier comment](https://github.com/guacml/guacml/pull/18#discussion_r131238728)? What do you think about either making this dependency on "guac need to run before creating the pipeline" more explicit or running it ourselves if it wasn't run yet?

File: guacml/models/random_forest.py (hunk 7 above)
Comment (hwartig) [reply guacml/models/random_forest.py:61]: Ok I'm good with leaving it as is until we have a better feeling / idea about it.

File: guacml/enums.py (hunk 1 above)
Comment (exviva) [reply guacml/enums.py:3]: I'm not such a big fan of those enums, we never validated by type anyway, and we already have 90% of other settings as strings, so I'd rather we add validation per setting. I agree we need input validation with clear error messages, but I'm not sure enums are the best way to achieve that, also seems like enums are not really that popular in the python world, most big libraries just use strings.

File: guacml/pipeline.py (hunk 10 above)
Comment (exviva) [reply guacml/pipeline.py:13]: Definitely good idea to add this check!

File: guacml/pipeline.py (hunk 9 above)
Comment (exviva) [reply guacml/pipeline.py:33]: That'd be really nice to have, but I'm a bit afraid of touching column analyzer right now, because it's pretty convoluted. But I definitely like the idea of having transformers, which are collected during `run()`, and then we can replay them on a different dataset.
